Tests/DatadogTests/Datadog/Core/FeatureTests.swift

```diff
@@ -68,8 +59,83 @@ class FeatureStorageTests: XCTestCase {
         )
     }
 
+    func testWhenArbitraryWriterIsUsedInParallelWithRegularWriter_thenAllDataIsWrittenSafely() {
+        // Given
+        let storage = FeatureStorage(
+            featureName: .mockAny(),
+            dataFormat: DataFormat(prefix: "", suffix: "", separator: "#"),
+            directories: temporaryFeatureDirectories,
+            eventMapper: nil,
+            commonDependencies: .mockWith(consentProvider: .init(initialConsent: .granted))
+        )
+
+        // When
+        // swiftlint:disable opening_brace
+        callConcurrently(
+            closures: [
+                { storage.writer.write(value: "regular write") },
+                { storage.arbitraryAuthorizedWriter.write(value: "arbitrary write") }
+            ],
+            iterations: 25
+        )
+        // swiftlint:enable opening_brace
+
+        // Then
+        let dataWritten = readAllAuthorizedDataWritten(to: storage)
+            .map { $0.utf8String }
+        XCTAssertEqual(dataWritten.filter { $0 == "\"regular write\"" }.count, 25)
+        XCTAssertEqual(dataWritten.filter { $0 == "\"arbitrary write\"" }.count, 25)
+    }
+
+    func testGivenStorageWithEventMapper_whenWrittingData_itMapsValuesWrittenWithBothRegularAndArbitraryWriters() {
+        class EventMapperMock: EventMapper {
+            func map<T: Encodable>(event: T) -> T? {
+                switch event as! String {
+                case "drop": return nil
+                case "secret": return "redact" as? T
+                default: return event
+                }
+            }
+        }
+
+        // Given
+        let storage = FeatureStorage(
+            featureName: .mockAny(),
+            dataFormat: DataFormat(prefix: "", suffix: "", separator: "#"),
+            directories: temporaryFeatureDirectories,
+            eventMapper: EventMapperMock(),
+            commonDependencies: .mockWith(consentProvider: .init(initialConsent: .granted))
+        )
+
+        // When
+        ["value", "secret", "drop"].forEach { event in
+            storage.writer.write(value: event)
+            storage.arbitraryAuthorizedWriter.write(value: event)
+        }
+
+        // Then
+        let dataWritten = readAllAuthorizedDataWritten(to: storage)
+            .map { $0.utf8String }
+        XCTAssertEqual(dataWritten, ["\"value\"", "\"value\"", "\"redact\"", "\"redact\""])
+    }
+
     // MARK: - Helpers
 
+    private func readAllAuthorizedDataWritten(to storage: FeatureStorage) -> [Data] {
+        var dataAuthorizedForUpload: [Data] = []
+
+        while true {
+            if let nextBatch = storage.reader.readNextBatch() {
```

Comment:
Is there any way the reader could have a bug and repeat some return value indefinitely for `readNextBatch()` thus creating an infinite loop?
Instead, could this `while` be bound by some max count expectation given by the calling test case? The same way we pass a timeout to async expectations.

Reply:
In theory, this should invoke a failure in reader's tests, but because we randomize tests order this may hang up first. I replaced it with limited for-loop 👌.